File: #186_Reverse_Words_in_a_String_II/Solution.java
public class Solution {
    public void reverseWords(char[] s) {
        if (s == null || s.length == 0) return;
        /* reverse the array */
        int len = s.length;
        int i = 0, j = len - 1;
        while (i <= j) {
        	char tmp = s[j];
        	s[j] = s[i];
        	s[i] = tmp;
        	i++;
        	j--;
        }
        /* reverse word part, skip space */
        i = 0; j = 0;
        while (i < len && j < len) {
        	while(j + 1 < len && s[j + 1] != ' ') j++;
        	int start = i; int end = j;
        	while (start <= end) {
        		char tmp = s[end];
        		s[end] = s[start];
        		s[start] = tmp;
        		start++;
        		end--;
        	}
        	i = j + 2;
        	j = i;
        }
    }

    public static void main(String[] args) {
    	char[] t = {'t', 'h', 'e', ' ', 's', 'k', 'y', ' ', 'i', 's', ' ', 'b', 'l', 'u', 'e'};
    	Solution sol = new Solution();
    	sol.reverseWords(t);
    	for (char c : t) {
    		System.out.print(c);
    	}
    	System.out.println();
    }
}

File: #186_Reverse_Words_in_a_String_II/Solution.py
class Solution(object):

    def reverseWords(self, s):
		if not s:
			return
		self.reverseList(s, 0, len(s) - 1)
		i, j = 0, 0
		while j < len(s):
			j += 1
			if j == len(s) or s[j] == ' ':  # checking index of j must comes first
				self.reverseList(s, i, j - 1)
				i = j + 1

    def reverseList(self, s, start, end):
    	if not s:
    		return
    	if start < 0 or end >= len(s):
    		return
    	i, j = start, end
    	while i <= j:
    		tmp = s[i]
    		s[i] = s[j]
    		s[j] = tmp
    		i += 1
    		j -= 1

sol = Solution()
t = list("tab bbc")
sol.reverseWords(t)
print t